File: backend/routers/channel.py
from fastapi import APIRouter, HTTPException, Depends
from pydantic import BaseModel
from typing import List, Optional
import sqlite3
from datetime import datetime

try:
    from ..db import get_conn
    from ..routers.auth import get_current_user
except ImportError:
    from backend.db import get_conn
    from backend.routers.auth import get_current_user

router = APIRouter(prefix="/api/channels", tags=["channels"])

# Pydantic Models
class ChannelResponse(BaseModel):
    id: int
    name: str
    description: Optional[str]
    created_at: str
    message_count: int

class MessageCreate(BaseModel):
    content: str

class MessageResponse(BaseModel):
    id: int
    channel_id: int
    user_id: int
    username: str
    user_avatar: Optional[str]
    content: str
    created_at: str

# Get all channels
@router.get("", response_model=List[ChannelResponse])
def get_channels():
    """Get list of all channels"""
    conn = get_conn()
    cur = conn.cursor()
    
    cur.execute("""
        SELECT id, name, description, created_at, message_count
        FROM channels
        ORDER BY id ASC
    """)
    
    channels = []
    for row in cur.fetchall():
        channels.append({
            "id": row[0],
            "name": row[1],
            "description": row[2],
            "created_at": row[3],
            "message_count": row[4]
        })
    
    conn.close()
    return channels

# Get channel by ID
@router.get("/{channel_id}", response_model=ChannelResponse)
def get_channel(channel_id: int):
    """Get channel details by ID"""
    conn = get_conn()
    cur = conn.cursor()
    
    cur.execute("""
        SELECT id, name, description, created_at, message_count
        FROM channels
        WHERE id = ?
    """, (channel_id,))
    
    row = cur.fetchone()
    conn.close()
    
    if not row:
        raise HTTPException(status_code=404, detail="Channel not found")
    
    return {
        "id": row[0],
        "name": row[1],
        "description": row[2],
        "created_at": row[3],
        "message_count": row[4]
    }

# Get messages in a channel
@router.get("/{channel_id}/messages", response_model=List[MessageResponse])
def get_messages(channel_id: int, page: int = 1, page_size: int = 50):
    """Get messages in a channel with pagination"""
    conn = get_conn()
    cur = conn.cursor()
    
    # Verify channel exists
    cur.execute("SELECT id FROM channels WHERE id = ?", (channel_id,))
    if not cur.fetchone():
        conn.close()
        raise HTTPException(status_code=404, detail="Channel not found")
    
    offset = (page - 1) * page_size
    
    cur.execute("""
        SELECT id, channel_id, user_id, username, user_avatar, content, created_at
        FROM messages
        WHERE channel_id = ?
        ORDER BY created_at DESC
        LIMIT ? OFFSET ?
    """, (channel_id, page_size, offset))
    
    messages = []
    for row in cur.fetchall():
        messages.append({
            "id": row[0],
            "channel_id": row[1],
            "user_id": row[2],
            "username": row[3],
            "user_avatar": row[4],
            "content": row[5],
            "created_at": row[6]
        })
    
    conn.close()
    return messages

# Post a message to a channel
@router.post("/{channel_id}/messages", response_model=MessageResponse)
def post_message(channel_id: int, message: MessageCreate, username: str = Depends(get_current_user)):
    """Post a new message to a channel (requires authentication)"""
    conn = get_conn()
    cur = conn.cursor()
    
    # Verify channel exists
    cur.execute("SELECT id FROM channels WHERE id = ?", (channel_id,))
    if not cur.fetchone():
        conn.close()
        raise HTTPException(status_code=404, detail="Channel not found")
    
    # Get user details
    cur.execute("SELECT id, username, nickname, avatar_url FROM users WHERE username = ?", (username,))
    user_row = cur.fetchone()
    
    if not user_row:
        conn.close()
        raise HTTPException(status_code=404, detail="User not found")
    
    user_id = user_row[0]
    user_username = user_row[1]
    nickname = user_row[2] or user_username
    avatar_url = user_row[3]
    
    # Insert message
    cur.execute("""
        INSERT INTO messages (channel_id, user_id, username, user_avatar, content)
        VALUES (?, ?, ?, ?, ?)
    """, (channel_id, user_id, nickname, avatar_url, message.content))
    
    message_id = cur.lastrowid
    
    # Update channel message count
    cur.execute("""
        UPDATE channels
        SET message_count = message_count + 1
        WHERE id = ?
    """, (channel_id,))
    
    conn.commit()
    
    # Get the created message
    cur.execute("""
        SELECT id, channel_id, user_id, username, user_avatar, content, created_at
        FROM messages
        WHERE id = ?
    """, (message_id,))
    
    row = cur.fetchone()
    conn.close()
    
    return {
        "id": row[0],
        "channel_id": row[1],
        "user_id": row[2],
        "username": row[3],
        "user_avatar": row[4],
        "content": row[5],
        "created_at": row[6]
    }

# Delete a message (author or admin only)
@router.delete("/messages/{message_id}")
def delete_message(message_id: int, username: str = Depends(get_current_user)):
    """Delete a message (author or admin only)"""
    conn = get_conn()
    cur = conn.cursor()
    
    # Get message details
    cur.execute("""
        SELECT channel_id, user_id, username
        FROM messages
        WHERE id = ?
    """, (message_id,))
    
    row = cur.fetchone()
    if not row:
        conn.close()
        raise HTTPException(status_code=404, detail="Message not found")
    
    channel_id, message_user_id, message_username = row
    
    # Check if user is author or admin
    cur.execute("SELECT username FROM users WHERE id = ?", (message_user_id,))
    author_row = cur.fetchone()
    
    if username != "witw" and username != (author_row[0] if author_row else None):
        conn.close()
        raise HTTPException(status_code=403, detail="Not authorized to delete this message")
    
    # Delete message
    cur.execute("DELETE FROM messages WHERE id = ?", (message_id,))
    
    # Update channel message count
    cur.execute("""
        UPDATE channels
        SET message_count = message_count - 1
        WHERE id = ?
    """, (channel_id,))
    
    conn.commit()
    conn.close()
    
    return {"ok": True, "message": "Message deleted"}

File: web/src/lib/channel.ts
import { getChannelDb, getUsersDb } from "./db";

type UserRow = { id?: number; username: string; nickname?: string; avatar_url?: string; is_bot?: boolean | number };

function getUserMapById(userIds: number[]) {
  const unique = Array.from(new Set(userIds.filter((id) => Number.isFinite(id))));
  const map = new Map<number, UserRow>();
  if (unique.length === 0) return map;
  const usersDb = getUsersDb();
  const placeholders = unique.map(() => "?").join(",");
  const rows = usersDb.prepare(`SELECT id, username, nickname, avatar_url, is_bot FROM users WHERE id IN (${placeholders})`).all(...unique) as any[];
  rows.forEach((row) => map.set(row.id, row));
  return map;
}

function getUserByUsername(username: string) {
  const usersDb = getUsersDb();
  return usersDb.prepare("SELECT id, username, nickname, avatar_url, is_bot FROM users WHERE username = ?").get(username) as any;
}

export function listServers() {
  const db = getChannelDb();
  return db.prepare("SELECT id, name, icon_url, owner_id, created_at FROM servers ORDER BY id ASC").all();
}

export function listCategories(serverId: number) {
  const db = getChannelDb();
  return db.prepare("SELECT id, server_id, name, position FROM categories WHERE server_id = ? ORDER BY position ASC").all(serverId);
}

export function listChannels(serverId?: number, categoryId?: number) {
  const db = getChannelDb();
  let sql = "SELECT id, server_id, category_id, name, description, type, message_count FROM channels";
  const params: any[] = [];

  if (serverId) {
    sql += " WHERE server_id = ?";
    params.push(serverId);
    if (categoryId) {
      sql += " AND category_id = ?";
      params.push(categoryId);
    }
  }

  sql += " ORDER BY position ASC, id ASC";
  return db.prepare(sql).all(...params);
}

export function getChannelById(id: number) {
  const db = getChannelDb();
  return db.prepare("SELECT id, name, description, created_at, message_count FROM channels WHERE id = ?").get(id);
}

export function listMessages(channelId: number, page = 1, pageSize = 50) {
  const db = getChannelDb();
  const offset = (page - 1) * pageSize;
  const rows = db.prepare(`
    SELECT id, channel_id, user_id, username, user_avatar, content, created_at
    FROM messages
    WHERE channel_id = ?
    ORDER BY created_at DESC
    LIMIT ? OFFSET ?
  `).all(channelId, pageSize, offset) as any[];

  const userMap = getUserMapById(rows.map((r) => r.user_id));
  rows.forEach((row) => {
    const user = row.user_id ? userMap.get(row.user_id) : null;
    row.user_avatar = user?.avatar_url || row.user_avatar || "";
    row.username = user?.nickname || row.username;
    row.is_bot = !!user?.is_bot;
  });

  return rows;
}

export function postMessage(channelId: number, username: string, content: string) {
  const db = getChannelDb();
  const user = getUserByUsername(username) as any;
  if (!user) return null;
  const nickname = user.nickname || user.username;
  db.prepare("INSERT INTO messages (channel_id, user_id, username, user_avatar, content) VALUES (?, ?, ?, ?, ?)")
    .run(channelId, user.id, nickname, user.avatar_url, content);
  db.prepare("UPDATE channels SET message_count = message_count + 1 WHERE id = ?").run(channelId);
  const row = db.prepare("SELECT id, channel_id, user_id, username, user_avatar, content, created_at FROM messages WHERE rowid = last_insert_rowid()").get();
  return row;
}

export function deleteMessage(messageId: number, username: string) {
  const db = getChannelDb();
  const row = db.prepare("SELECT channel_id, user_id, username FROM messages WHERE id = ?").get(messageId) as any;
  if (!row) return { ok: false, error: "not_found" };
  const user = row.user_id ? getUsersDb().prepare("SELECT username FROM users WHERE id = ?").get(row.user_id) as any : null;
  const isAdmin = username === "witw";
  if (!isAdmin && username !== (user?.username || row.username)) {
    return { ok: false, error: "forbidden" };
  }
  db.prepare("DELETE FROM messages WHERE id = ?").run(messageId);
  db.prepare("UPDATE channels SET message_count = message_count - 1 WHERE id = ?").run(row.channel_id);
  return { ok: true };
}

export function listServerMembers(serverId: number) {
  const db = getChannelDb();
  const members = db.prepare("SELECT user_id, nickname, joined_at FROM server_members WHERE server_id = ?").all(serverId) as any[];
  if (members.length === 0) return [];

  const userMap = getUserMapById(members.map(m => m.user_id));
  return members.map(m => {
    const user = userMap.get(m.user_id);
    return {
      user_id: m.user_id,
      username: user?.username || "Unknown",
      nickname: m.nickname || user?.nickname || user?.username || "Unknown",
      avatar_url: user?.avatar_url || "",
      joined_at: m.joined_at
    };
  });
}

export function createChannel(name: string, description = "", serverId?: number, categoryId?: number) {
  const db = getChannelDb();
  const cleanName = String(name || "").trim();
  if (!cleanName) return { ok: false, error: "name_required" };

  // Check if exists in the same server
  const exists = db.prepare("SELECT id FROM channels WHERE name = ? AND server_id = ?").get(cleanName, serverId);
  if (exists) return { ok: false, error: "exists" };

  db.prepare("INSERT INTO channels (name, description, server_id, category_id, message_count) VALUES (?, ?, ?, ?, 0)")
    .run(cleanName, description || "", serverId || null, categoryId || null);

  const row = db.prepare("SELECT id, name, description, server_id, category_id, created_at, message_count FROM channels WHERE rowid = last_insert_rowid()").get();
  return { ok: true, channel: row };
}
export function createServer(name: string, ownerId: number, iconUrl?: string) {
  const db = getChannelDb();
  const cleanName = String(name || "").trim();
  if (!cleanName) return { ok: false, error: "name_required" };

  try {
    const res = db.prepare("INSERT INTO servers (name, owner_id, icon_url) VALUES (?, ?, ?)")
      .run(cleanName, ownerId, iconUrl || null);
    const serverId = res.lastInsertRowid as number;

    // 1. Create default categories
    const catTextRes = db.prepare("INSERT INTO categories (server_id, name, position) VALUES (?, ?, ?)")
      .run(serverId, "文字频道", 1);
    db.prepare("INSERT INTO categories (server_id, name, position) VALUES (?, ?, ?)")
      .run(serverId, "语音频道", 2);

    // 2. Create initial channel in the first category
    db.prepare("INSERT INTO channels (name, server_id, category_id, type) VALUES (?, ?, ?, ?)")
      .run("综合", serverId, catTextRes.lastInsertRowid, "text");

    // 3. Auto-join owner as a member
    db.prepare("INSERT OR IGNORE INTO server_members (server_id, user_id, nickname) VALUES (?, ?, ?)")
      .run(serverId, ownerId, null);

    const row = db.prepare("SELECT id, name, icon_url, owner_id, created_at FROM servers WHERE id = ?").get(serverId);
    return { ok: true, server: row };
  } catch (err: any) {
    if (err.message?.includes("UNIQUE constraint failed")) {
      return { ok: false, error: "exists" };
    }
    throw err;
  }
}
